whisper.cpp

```diff
@@ -2432,34 +2431,43 @@ static void log_mel_spectrogram_worker_thread(int ith, const std::vector<float>
                                               const whisper_filters &filters, bool speed_up, whisper_mel &mel) {
     std::vector<float> fft_in(fft_size, 0.0);
     std::vector<float> fft_out(2 * fft_size);
+    // Is using 32-bit float to calculate log_mel appropriate?
+    // 32-bit float has about 7 digits of precision, but minimum value of log_mel is 1e-10.
     int n_fft = 1 + (speed_up ? fft_size / 4 : fft_size / 2);
+    int i = ith;
 
-    for (int i = ith; i < mel.n_len; i += n_threads) {
+    // Calculate FFT only when fft_in are not all zero
+    for (; i < std::min((n_samples / fft_step) + 1, mel.n_len); i += n_threads) {
         const int offset = i * fft_step;
 
-        // apply Hanning window
-        for (int j = 0; j < fft_size; j++) {
-            if (offset + j < n_samples) {
-                fft_in[j] = hann[j] * samples[offset + j];
-            } else {
-                fft_in[j] = 0.0;
-            }
+        // apply Hanning window (~10% faster)
+        for (int j = 0; j < std::min(fft_size, n_samples - offset); j++) {
+            fft_in[j] = hann[j] * samples[offset + j];
+        }
+        // fill the rest with zeros
+        if (n_samples - offset < fft_size) {
+            std::fill(fft_in.begin() + (n_samples - offset), fft_in.end(), 0.0);
         }
 
-        // FFT -> mag^2
+        // FFT
         fft(fft_in, fft_out);
 
+        // Calculate modulus^2 of complex numbers
+        // Use pow(fft_out[2 * j + 0], 2) + pow(fft_out[2 * j + 1], 2) causes inference quality problem? Interesting.
         for (int j = 0; j < fft_size; j++) {
             fft_out[j] = (fft_out[2 * j + 0] * fft_out[2 * j + 0] + fft_out[2 * j + 1] * fft_out[2 * j + 1]);
         }
-        for (int j = 1; j < fft_size / 2; j++) {
-            fft_out[j] += fft_out[fft_size - j];
+
+        // The frequency spectrum produced by real input data is symmetrical around the Nyquist frequency.
+        // This is where the actual issue lies
+        for (int j = 0; j < fft_size / 2; j++) {
+            fft_out[j] = (fft_out[fft_size - j - 1] + fft_out[j + 1]) / 2;
         }
```

Comment:
Help me out a bit here - it's very unclear to me what this code is here for.  We get the symmetrical spectrum in `fft_out[0...fft_size]` for free from the previous loop.  Not only that, but inference seems to work fine with it commented out entirely.  What's it supposed to be doing?

Reply:
Your thinking is correct. Theoretically, this step is not needed at all. Because the FFT_SIZE is `400`, `FFT_OUT[200]` is the Nyquist Component; `FFT_OUT[0]` is the DC Component, which is the average of all 400 samples in FFT_IN. The remaining parts, `FFT_OUT[1...199]` and `FFT_OUT[201...399]`, are symmetrical. I guess the original intention of the code was just to make the result more accurate, so the symmetrical parts were added together? But this is obviously unreasonable, because the amplitude of `FFT_OUT[0]` does not match the rest. So I made a modification, turning it into adding them together and then dividing by 2. However, I made a guess that perhaps `FFT_OUT[0]`, as the DC Component, has no actual meaning? So I shifted the whole thing forward by one position, and the effect was surprisingly good, so I kept it. But I don't have enough evidence to support this guess yet.

> 

https://github.com/ggerganov/whisper.cpp/blob/a4bb2df36aeb4e6cfb0c1ca9fbcf749ef39cc852/whisper.cpp#L2455-L2457

Reply:
I think the most important thing is to establish a complete and scientific WER (Word Error Rate) detection framework, so that we can quantitatively compare the quality. Otherwise, it would be too subjective.

Reply:
There are three things to play off here:

1. The word error rate
2. Byte-for-byte matching OpenAI's mel implementation choice
3. Theoretical approach at this step in particular

While 1 is what we're after, and using that as a backstop is undoubtedly going to pay off, it feels like 2 would be a very useful stopping-off point to check, so we know we're at least putting the same things into the black box as they are; and 3 would help us avoid bugs on the way there.

What worries me here is this:

> I guess the original intention of the code was just to make the result more accurate, so the symmetrical parts were added together?

If the things being averaged weren't already identical numbers, then wouldn't that point to a bug in the FFT implementation?  Averaging them should always have been a no-op.

> However, I made a guess that perhaps FFT_OUT[0], as the DC Component, has no actual meaning?

That's not *quite* true.  `fft_out[0]` is referred to as the "dc component" as a shorthand, but remember that it's actually more accurately the bin which happens to centre at 0Hz. It contains information for all frequencies from 0 up to the half the bin width. With 400 samples at 16kHz the bin width is 40Hz, so `fft_out[0]` has everything up to 20Hz in it - and also it gets any signal power at the sample frequency aliased into it.  I don't think I'd make a bet that neither of those things matters.  I know human vocal chords don't typically go that low, but I wouldn't write off there being transient phoneme information in the DC component.

Reply:
Regarding how to measure the word error rate, I found that there are many well-annotated datasets online, such as Mozilla's Common Voice. An article is divided into many sentences, read by different users, ensuring a diversity of accents, and including not just English but many other languages. The dataset is under CC-0, so there won't be any potential legal issues. Therefore, we can write a script in Python to batch test the WER of whisper.cpp as well as OpenAI's whisper.

Reply:
> While 1 is what we're after, and using that as a backstop is undoubtedly going to pay off, it feels like 2 would be a very useful stopping-off point to check, so we know we're at least putting the same things into the black box as they are; and 3 would help us avoid bugs on the way there.

Agree. If needed, match the `whisper.cpp` precision so that we have identical input with PyTorch.
Before doing any kind of WER, we need identical input to the transformer and I know it is not the case currently. Not sure if it is just a matter of precision or if there is a bug in `whisper.cpp` - needs similar analysis to what @bobqianic did in the comment above.

> If the things being averaged weren't already identical numbers, then wouldn't that point to a bug in the FFT implementation? Averaging them should always have been a no-op.

Yes, but I believe the values are identical and this step is basically a noop. Needs confirmation. I guess I've put it there due to some doubts I had back when working on `ggwave` and was new to audio processing, but indeed the spectrum should be symmetrical.

Don't think we need to discard `fft_out[0]`. Again - let's compare what PyTorch is doing to make sure.

Reply:
> Help me out a bit here - it's very unclear to me what this code is here for. We get the symmetrical spectrum in `fft_out[0...fft_size]` for free from the previous loop. Not only that, but inference seems to work fine with it commented out entirely. What's it supposed to be doing?

### Confirmed. This step is unnecessary.  
I carefully examined the C++ implementation of PyTorch's STFT, and it directly removes all the bins after the Nyquist bin.

Reply:
Good to know I'm not going completely mad :-) If you have a look at [@lunixbochs' pocketfft branch](https://github.com/ggerganov/whisper.cpp/pull/583/files#diff-7b638c730017b16cf40fc5b8d402c8729b9d4cce513b1d1f4d2e07c2e62ed9b4R2254) from a little while back you'll see that the real to complex pocketfft FFT implementation gives you back a buffer of length `(fft_size/2)+1`.  Using `std::complex` directly might clarify things, even without swapping out the FFT implementation.